### PR TITLE
Manage session state passed between client and server properly.

### PR DESCRIPTION
**Pull request by vivekmenezes**

### CODE REVIEW

File: sql/server.go [hunk 1]
```diff
@@ -153,37 +153,45 @@ func (p parameters) Arg(i int) (parser.Datum, bool) {
...
-func (s *Server) exec(req driver.Request) (driver.Response, error) {
-	var resp driver.Response
+func (s *Server) exec(req driver.Request) (resp driver.Response, err error) {
 
 	// Pick up current session state.
 	// The request user is validated in ServeHTTP. Even in insecure mode,
 	// it is guaranteed not to be empty.
-	planner := planner{user: req.GetUser()}
+	planMaker := planner{user: req.GetUser()}
+	defer func(p *planner) {
```
Comment (tamird) [inline sql/server.go]: I'm a little uncomfortable with this pattern. perhaps we could break `exec` into two methods, and have one execute the inner logic while the outer takes care to set the session?

**Comment by tamird** [comment]
Presumably this code didn't work correctly before this change. Could you include some tests that demonstrate it now does the right thing?

**Comment by vivekmenezes** [comment]
I added a test. It does look a bit forced because we currently only store the database in the settings. Once the transaction state is also in the settings this change will be more valuable.

File: sql/server.go (hunk 1 above)
Comment (vivekmenezes) [inline sql/server.go]: We currently already have two methods ServeHTTP and exec(), where an error found in ServeHTTP doesn't result in session state being set by the server. But I do agree the use of defer makes one have to look at all the error scenarios and wonder if they all should legitimately allow session to be sent back.

File: sql/testdata/set [hunk 2]
```diff
@@ -29,3 +29,17 @@ SET DATABASE = ""
...
+
+# SET succeeds
+query T colnames
+SHOW TABLES from foo
+----
+Table
+bar
+
+
+
```
Comment (petermattis) [inline sql/testdata/set]: Delete the spurious blank lines at the end of this file.

File: sql/testdata/set [hunk 3]
```diff
@@ -29,3 +29,17 @@ SET DATABASE = ""
 
 statement error no database specified
 SHOW TABLES
+
+# SET statement succeeds, CREATE TABLE fails.
+statement error table \"bar\" already exists
+SET DATABASE = foo;CREATE TABLE bar (k INT PRIMARY KEY)
```
Comment (petermattis) [inline sql/testdata/set]: `s/foo;CREATE/foo; CREATE/g`

File: sql/server.go [hunk 4]
```diff
@@ -153,37 +153,45 @@ func (p parameters) Arg(i int) (parser.Datum, bool) {
...
+		// Update session state.
+		var marshallErr error
+		if resp.Session, marshallErr = gogoproto.Marshal(&(p.session)); marshallErr != nil {
```
Comment (petermattis) [inline sql/server.go]: `s/&(p.session)/&p.session/g`

File: sql/server.go (hunk 1 above)
Comment (tamird) [inline sql/server.go]: > But I do agree the use of defer makes one have to look at all the error scenarios and wonder if they all should legitimately allow session to be sent back.

Then, should that change be made here?

File: sql/server.go (hunk 4 above)
Comment (vivekmenezes) [inline sql/server.go]: done

File: sql/testdata/set (hunk 2 above)
Comment (vivekmenezes) [inline sql/testdata/set]: done

File: sql/server.go (hunk 1 above)
Comment (vivekmenezes) [inline sql/server.go]: yeah this is the correct behavior. It's just that for the moment correcting it only fixes a minor issue. But the way it's currently implemented is erroneous. Thanks.

File: sql/driver/conn.go [hunk 5]
```diff
@@ -77,8 +77,12 @@ func (c *conn) Query(stmt string, args []driver.Value) (*rows, error) {
 		}
 		params = append(params, param)
 	}
+	// Forget the session state, and use the one provided in the server
```
Comment (tamird) [inline sql/driver/conn.go]: how come the "forgetting" part is necessary?

File: sql/server.go [hunk 6]
```diff
@@ -153,37 +153,45 @@ func (p parameters) Arg(i int) (parser.Datum, bool) {
...
 	// it is guaranteed not to be empty.
-	planner := planner{user: req.GetUser()}
+	planMaker := planner{user: req.GetUser()}
+	defer func(p *planner) {
+		// Update session state.
+		var marshallErr error
+		if resp.Session, marshallErr = gogoproto.Marshal(&p.session); marshallErr != nil {
+			err = marshallErr
+		}
+	}(&planMaker)
```
Comment (tamird) [inline sql/server.go]: this doesn't need to take an argument; it can simply close over the planner

File: sql/driver/conn.go (hunk 5 above)
Comment (vivekmenezes) [inline sql/driver/conn.go]: On Tue, Aug 18, 2015, 10:03 PM Tamir Duberstein <notifications@github.com>
wrote:

In sql/driver/conn.go
<https://github.com/cockroachdb/cockroach/pull/2150#discussion_r37373867>:

> @@ -77,8 +77,12 @@ func (c *conn) Query(stmt string, args []driver.Value)
(*rows, error) { > } > params = append(params, param) > } > + // Forget the
session state, and use the one provided in the server

how come the "forgetting" part is necessary?

The forgetting part is not necessary. But I think what makes sense here is
for me to add a panic on the session update side to check that the client
session is nil.


—
Reply to this email directly or view it on GitHub
<https://github.com/cockroachdb/cockroach/pull/2150/files#r37373867>.

**Comment by petermattis** [comment]
LGTM

File: sql/server.go (hunk 6 above)
Comment (vivekmenezes) [inline sql/server.go]: done. Turns out that does work. I thought the context works like function parameters, but it actually can change under the covers with defer using the freshest value of planMaker. Thanks

**Pull request merged**